src/Desktop.Analyzers/Core/TypesShouldNotExtendCertainBaseTypes.cs

```diff
@@ -15,95 +18,63 @@ public abstract class TypesShouldNotExtendCertainBaseTypesAnalyzer : DiagnosticA
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private static readonly Dictionary<string, string> s_badBaseTypesToMessage = new Dictionary<string, string>
+                                                    {
+                                                        {"System.ApplicationException", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemApplicationException},
+                                                        {"System.Xml.XmlDocument", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemXmlXmlDocument},
+                                                        {"System.Collections.CollectionBase", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemCollectionsCollectionBase},
+                                                        {"System.Collections.DictionaryBase", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemCollectionsDictionaryBase},
+                                                        {"System.Collections.Queue", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemCollectionsQueue},
+                                                        {"System.Collections.ReadOnlyCollectionBase", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemCollectionsReadOnlyCollectionBase},
+                                                        {"System.Collections.SortedList", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemCollectionsSortedList},
+                                                        {"System.Collections.Stack", DesktopAnalyzersResources.TypesShouldNotExtendCertainBaseTypesMessageSystemCollectionsStack},
+                                                    };
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var badBaseTypes = s_badBaseTypesToMessage.Keys
+                                .Select(bt => context.Compilation.GetTypeByMetadataName(bt))
+                                .Where(bt => bt != null)
+                                .ToImmutableHashSet();
+
+            if (badBaseTypes.Count > 0)
+            {
+                context.RegisterSymbolAction((saContext) =>
+                    {
+                        var namedTypeSymbol = saContext.Symbol as INamedTypeSymbol;
+
+                        var containedBadBaseType = badBaseTypes.Where(bbt => bbt.Equals(namedTypeSymbol.BaseType)).FirstOrDefault();
```

Comment:
you have hashset. I think you can do badBaseTypes.Contains(namedTypeSymbol.BaseType)